Line info in tokens

elijah-ezaga opened the pull request:
a) Tokens now have line number info added
b) Heatfiler can integrate with external adapter who want to save report in various formats.

elijah-ezaga committed:
a) Added line information into tokens and also file statistics
b) Temporary fix for null error when dealing with 'if' token
c) Changed attribute to indicate script profiling to 'heat-filer-type'
d) Error reporting when saving into local storage
e) Can now generate reports from Heatfiler through external adapter.

elijah-ezaga committed:
Added token (start) offset into file stats object for coverage

elijah-ezaga committed:
a) Fixed line number bug for tokens
b) Added Qunit tests for minified and unminified Javascript code

qfox commented on lib/tok.js:
```diff
@@ -342,13 +344,14 @@
         var type = this.nextWhiteToken(expressionStart);
         if (toStream) {
-          var token = {type:type, value:this.getLastValue(), start:this.lastStart, stop:this.pos, white:this.tokens.length};
+          var lastValue = this.getLastValue();
```
Since the last value is assigned to the `token`, you don't really need this since you're also passing on the token itself. You can simply read `token.value` to get these same value. This saves you the argument and extra changes here.

qfox commented on lib/tok.js:
```diff
@@ -362,6 +365,26 @@
+     * comment token.
+     * */
+    addLineCount: function(token, type, lastValue){
+      token['line'] = this.lineCounter;
```
Are you doing `['line']` to prevent mangling? I think it's fine to use `token.line` here.

qfox commented on lib/tok.js:
```diff
@@ -362,6 +365,26 @@
+        return;
+      }
+
+      var carriageMatches = lastValue.match(/[\r\n]/g);
```
I think you missed the exotic PS and LS chars (`0x2028` and `0x2029`) which are formally considered newlines in JS.

How would this count multi-line comments with multiple newlines? And before you address this please consider that JS considers the combo `crlf` as a single newline for the sake of line reporting.

elijah-ezaga replied:
Yes I would now have to change it. My understanding was that for instance n lines would have at least n-1 line separators.

qfox commented on src/heatfiler.js:
```diff
@@ -413,7 +442,9 @@
         this.exposeGlobals(true);
         this.toLocalStorage('meta');
-        this.run(this.profiledFidMap[0]);
+        this.profiledFidMap.forEach(function(fid){
```
Was this a bug? If so, thanks! :)

elijah-ezaga replied:
Yes :)

qfox commented on src/transformer.js:
```diff
@@ -179,7 +179,17 @@
+    /** Adds line number info to file statistics */
+    addLineInfoToStat: function(statInfo, token){
+      if('line' in token){
+        statInfo['line'] = token['line'];
```
Please change to `.line` unless there's a reason for using dynamic properties this way.

qfox commented on src/transformer.js:
```diff
@@ -179,7 +179,17 @@
+    addLineInfoToStat: function(statInfo, token){
+      if('line' in token){
+        statInfo['line'] = token['line'];
+        statInfo['start'] = token['start'];
```
(Same for `.start`)

qfox commented on src/transformer.js:
```diff
@@ -295,11 +309,14 @@
-          token.ownerCountType = transformer.typeIf;
-          token.ownerCountFuncId = token.ownerFuncToken.white;
-          func.ifs.push(0);
+          //TODO For what case will 'func' become undefined? Pls investigate.
```
This implies you've found a case where this happens. Can you file a ticket and report the input used to get that problem?

elijah-ezaga replied:
Done. #17

qfox replied:
Thanks. This issue is now fixed. Please rebase on top of master to make sure some of your other concerns have been addressed now.

pvdz reviewed (commented):
Thanks for the PR! 

It looks like a good addition but please address the questions raised :)

Some friendly feedback: Don't make changes that aren't necessary (like adding empty lines or the `.gitignore`) or explain why you are adding them. This helps the reviewer because less noise is better.

qfox commented:
FYI: I'm going to rebase the branch and apply these changes myself. I don't want to spend too much time on this at the moment but also don't want to leave it open.

qfox commented:
Hm. Okay that's not how github works. Alright, then that's up to you.

elijah-ezaga commented:
Really displeased I've been quite occupied and could not continue. I do the rebase and fix the noted issues.

Commit:
a) Removed new lines previously added
b) Used dot notation for object property access
c) Changed next() function removing redundant lastValue variable

Commit:
a) Used dot notation for writing to object
b) Adjusted regex for counting lines.

elijah-ezaga commented:
All changes made and ready for merging.

qfox commented:
Thank you :)
